server/script/migrateAssetsLocation.js

```diff
@@ -0,0 +1,194 @@
+  });
+}
+
+async function migrateActivity(activity) {
+  const { repositoryId, data: meta } = activity;
+  const fileMetas = getActivityMetadata(activity)
+    .filter(it => it.type === 'FILE')
+    .map(it => it.key);
+  const data = await getNewMeta(fileMetas, meta, repositoryId);
+  return { data };
+}
+
+async function migrateContentElements(transaction) {
+  const contentElements = await ContentElement.findAll({
+    where: { type: { [Op.in]: ceTypes } },
+    transaction
+  });
+  return Promise.each(contentElements, async it => {
+    const payload = await migrateContentElement(it);
+    return it.update(payload, { transaction });
+  });
+}
+
+async function migrateContentElement(element) {
+  const data = await migrateContentElementData(element);
+  const meta = await migrateContentElementMeta(element);
+  return { data, meta };
+}
+
+function migrateContentElementData(element) {
+  return invokeAction(element.type)(element);
+}
+
+async function migrateContentElementMeta(element) {
+  const { repositoryId } = element;
+  const metaInputs = schemasIds
+    .map(id => getElementMetadata(id, element))
+    .filter(meta => !meta.isEmpty)
+    .map(meta => meta.inputs);
+  const fileMetas = flatten(metaInputs)
+    .filter(it => it.type === 'FILE')
+    .map(it => it.key);
+  return getNewMeta(fileMetas, element.meta, repositoryId);
+}
+
+async function migrateRevisions(transaction) {
+  const revisions = await Revision.findAll({
+    where: { entity: { [Op.in]: revisionsTypes } },
+    transaction
+  });
+  return Promise.each(revisions, async it => {
+    const payload = await migrateRevision(it);
+    return it.update(payload, { transaction });
+  });
+}
+
+async function migrateRevision(revision) {
+  const { entity, state } = revision;
+  const payload = await (mapEntityToAction[entity] && mapEntityToAction[entity](state));
+  return { state: { ...state, ...payload } };
+}
+
+async function getNewMeta(fileMetas, meta, repositoryId) {
+  const newMeta = await Promise.map(toPairs(meta), async it => {
+    const [id, value] = it;
+    if (!fileMetas.includes(id)) return it;
+    const url = get(value, 'url');
+    if (!url) return it;
+    const { key, newKey } = getKeysFromUrl(url, repositoryId) || {};
+    if (!key || !newKey) return it;
+    await storage.copyFile(key, newKey);
+    return [id, {
+      ...value,
+      key: newKey,
+      url: `${protocol}${newKey}`,
+      publicUrl: await storage.getFileUrl(newKey)
+    }];
+  });
+  return fromPairs(newMeta);
+}
+
+async function imageMigrationHandler(element) {
+  const { repositoryId, data } = element;
+  const url = get(data, 'url');
+  if (!url) return data;
+  const { key, newKey } = getKeysFromUrl(url, repositoryId) || {};
+  if (!key || !newKey) return data;
+  await storage.copyFile(key, newKey);
+  return { ...element.data, url: newKey };
+}
+
+async function embedsMigrationHandler(element) {
+  const { repositoryId, data } = element;
+  const embeds = await Promise.map(
```

Comment:
Hm..maybe extract this map logic as `getEmbeds` separate function. It's lot going one, so it will be more readable

Reply:
```js
async ([id, embed]) => {
  if (embed.type !== 'VIDEO') return [id, embed];
  const data = await defaultMigrationHandler({ repositoryId, ...embed });
  return [id, { ...embed, data }];
}
```
The map iteration function is not that heavy (3 lines of code). What do you think of extracting full get embeds logic(map + reduce) instead?

```js
const embeds = getEmbeds(...);
return { ...data, embeds };

function getEmbeds(...) {
  return Promise.map(
    Object.entries(embeds),
    async ([id, embed]) => {
      if (embed.type !== 'VIDEO') return [id, embed];
      const data = await defaultMigrationHandler({ repositoryId, ...embed });
      return [id, { ...embed, data }];
    })
    .reduce((acc, [id, embed]) => ({ ...acc, [id]: embed }), {});
}

Reply:
That was actually what I meant you to do 😉